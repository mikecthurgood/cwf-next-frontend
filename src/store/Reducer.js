const reducer = (state, action) => {
    switch (action.type) {
        
        case 'setFilterSelection':
            return { ...state, filterSelection: action.data}
        case 'setLoginError':
            return { ...state, loginError: action.data}
        case 'setLoginMenuVisible':
            return { ...state, loginMenuVisible: action.data}
        case 'setMobileMenuVisible':
            return { ...state, mobileMenuVisible: action.data}
        case 'setSearchBarVisibility':
            return { ...state, searchBarVisible: action.data}
        case 'setSearchFilter':
            return { ...state, searchFilter: action.data}
        case 'setScrollPosition':
            return { ...state, scrollPosition: action.data}
        case 'setSignUpSuccess':
            return { ...state, signUpSuccess: action.data}
        case 'setSingleWall':
            return { ...state, singleWall: action.data}
        case 'setSortInputVisibility':
            return { ...state, sortInputVisible: action.data}
        case 'setUser':
            return { ...state, user: action.data}
        case 'setPostCode':
            return { ...state, userPostCode: action.data}
        case 'setWalls':
            return { ...state, walls: action.data}
        case 'signOut' :
            localStorage.removeItem('token')
            localStorage.removeItem('userName')
            localStorage.removeItem('userId')
            return { 
                ...state,
                user: {username: null, userId: null, isAuth: false, token: ''},
                mobileMenuVisible: false,
                loginMenuVisible: false
            }
        default: 
            throw new Error('Invalid action type dispatched')
    }
}
export default reducer