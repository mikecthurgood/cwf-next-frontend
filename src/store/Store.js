import React, { useReducer } from  'react';
import reducer from './Reducer'

const initialState = {
    filterSelection: [],
    loginError: false,
    loginMenuVisible: false,
    mobileMenuVisible: false,
    searchBarVisible: false,
    scrollPosition: 0,
    signUpSuccess: false,
    singleWall: {},
    sortInputVisible: false,
    user: { username: null, userId: null, isAuth: false, token: '' },
    userPostCode: '',
    walls: [],
}

const State = React.createContext(initialState)
const Dispatch = React.createContext(null)

const Store = ({children}) => {
    
    const [state, dispatch] = useReducer(reducer, initialState)

    return (
        <Dispatch.Provider value={dispatch} >
            <State.Provider value={state} >
                {children}
            </State.Provider>
        </Dispatch.Provider>
    )
}


export { State, Dispatch, Store }

