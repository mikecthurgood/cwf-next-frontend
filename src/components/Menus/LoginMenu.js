import React, { useState, useContext } from 'react'
import { Link } from 'react-router-dom';
import { Dispatch, State } from '../../context/Store';
import './loginMenu.scss'

const AuthForm = React.lazy(() => import('../form/AuthForm'))

const LoginMenu = ({loginHandler, signupHandler}) => {
    const { loginError, loginMenuToggle, user, loginMenuVisible } = useContext(State)
    const dispatch = useContext(Dispatch)
    const [loginVisible, setLoginVisible] = useState(false)

    function loginToggle () {
        setLoginVisible(!loginVisible)
    }

    function signOut () {
        dispatch({type: 'signOut'})
    }

    return (
        <div className={`login__menu-container ${loginMenuVisible ? 'visible' : ''} ${user.username ? '' : 'logged-out'}`}>
            <div className='login__menu-login-register'>
                {user && user.username ? 
                <h5 onClick={signOut}>Log out</h5>
                :
                (<>
                    <h5 onClick={loginToggle}>Login</h5>
                    <div className={`login__menu-auth__form ${loginVisible ? 'visible' : ''}`}>
                        <AuthForm 
                            loginError={loginError}
                            signup={false} 
                            hideRegisterButton={true}
                            user={false}
                            loginToggle={loginToggle}
                            LoginMenuToggle={loginMenuToggle}
                            loginHandler={loginHandler}
                            signupHandler={signupHandler}
                        />
                    </div>
                    <Link to='/signup'><h5 onClick={loginMenuToggle}>Register</h5></Link>
                </>)
                }
            </div>
        </div>
    )
}

export default LoginMenu