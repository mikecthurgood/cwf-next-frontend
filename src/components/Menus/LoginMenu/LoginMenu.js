import React, { useState, useContext } from 'react'
import Styled from 'styled-components';
import { Dispatch, State } from '../../../store/Store';
import AuthForm from '../../Form/Login';

const LoginMenu = () => {
    const { user, loginMenuVisible } = useContext(State)
    const dispatch = useContext(Dispatch)
    const [loginVisible, setLoginVisible] = useState(false)

    function loginMenuToggle () {
        dispatch({type: 'setLoginMenuVisible', data: !loginMenuVisible})
    }
    
    function loginFormToggle () {
        setLoginVisible(!loginVisible)
    }

    function signOut () {
        dispatch({type: 'signOut'})
    }

    return (
        <LoginMenuContainer className={`${loginMenuVisible ? 'visible' : ''} ${user.username ? '' : 'logged-out'}`}>
            <div className='login__menu-login-register'>
                {user && user.username ? 
                <h5 onClick={signOut}>Log out</h5>
                :
                (<>
                    <h5 onClick={loginFormToggle} data-testid='login-form-control'>Login</h5>
                    <AuthFormContainer className={loginVisible ? 'visible' : ''}>
                        <AuthForm layout={'compact'} />
                    </AuthFormContainer>
                    <a href='/signup'><h5 onClick={loginMenuToggle}>Register</h5></a>
                </>)
                }
            </div>
        </LoginMenuContainer>
    )
}

export default LoginMenu

const LoginMenuContainer = Styled.div`
    position: absolute;
    top: 0;
    right: 0;
    display: flex;
    z-index: 999;
    width: 50%;
    max-width: 220px;
    background-color: #282c34;
    color: white;
    padding: 44px 20px 20px 20px;
    opacity: 0.95;
    font-size: 16px;
    overflow-y: hidden;
    transform: translateY(-300px);
    transition: transform 0.45s ease-out; 
    &.visible {
        transform: translateY(0px);
        transition: transform 0.45s ease-out; 
    }

    h5, Link, a {
        padding: 0;
        margin: 20px 0 0 0;
        font-size: 16px;
        text-align: left;
        color: white;
        text-decoration: none;
        text-align: center;
        cursor: pointer;
    }

.login__menu-login-register {
    width: 100%;
}

@media (min-width: 768px) {
  display: none;
}
`

const AuthFormContainer = Styled.div`
    position: relative;
    max-height: 0px;
    overflow-y: hidden;
    transition: max-height 0.25s ease-out; 
    width: 100%;
    &.visible {
        max-height: 300px;
        transition: max-height 0.45s ease-in; 
    }

`