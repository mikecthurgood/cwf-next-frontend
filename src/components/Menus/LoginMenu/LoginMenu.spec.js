import React from 'react';
import { cleanup, render, fireEvent } from '@testing-library/react';
import { Dispatch, State } from '../../../store/Store';
import LoginMenu from '.';

describe('Login Menu', () => {
    afterEach(cleanup);
    it('render the component', () => {
        const dispatch = jest.fn();
        const state = {
            user: { username: 'Mr Test' },
            loginMenuVisible: true
        }
        const { container } = render(
            <Dispatch.Provider value={dispatch}>
                <State.Provider value={state} >
                    <LoginMenu />
                </State.Provider>
            </Dispatch.Provider>,
        );

        expect(container).toMatchSnapshot()
    })

    it('should send a dispatch call to sign out a user if logged in', () => {
        const dispatch = jest.fn();
        const state = {
            user: { username: 'Mr Test' },
            loginMenuVisible: true
        }
        const { getByText } = render(
            <Dispatch.Provider value={dispatch}>
                <State.Provider value={state} >
                    <LoginMenu />
                </State.Provider>
            </Dispatch.Provider>,
        );

        fireEvent.click(getByText('Log out'));
        expect(dispatch).toHaveBeenCalledWith({type: 'signOut'})

    })

    it('should open the login form in the login menu', () => {
        const dispatch = jest.fn();
        const state = {
            user: { username: '' },
            loginMenuVisible: true
        }
        const { container, getByTestId } = render(
            <Dispatch.Provider value={dispatch}>
                <State.Provider value={state} >
                    <LoginMenu />
                </State.Provider>
            </Dispatch.Provider>,
        );

        fireEvent.click(getByTestId('login-form-control'));
        expect(container).toMatchSnapshot();
    })

    it('should send a user to the registration page', () => {
        const dispatch = jest.fn();
        const state = {
            user: { username: '' },
            loginMenuVisible: true
        }
        const { getByText } = render(
            <Dispatch.Provider value={dispatch}>
                <State.Provider value={state} >
                    <LoginMenu />
                </State.Provider>
            </Dispatch.Provider>,
        );

        fireEvent.click(getByText('Register'));
        expect(dispatch).toHaveBeenCalledWith({type: 'setLoginMenuVisible', data: false});
    })
});
