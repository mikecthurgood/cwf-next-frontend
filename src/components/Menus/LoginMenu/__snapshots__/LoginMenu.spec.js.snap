// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Menu render the component 1`] = `
<div>
  <div
    class="LoginMenu__LoginMenuContainer-sc-1ec36x2-0 guOatf visible "
  >
    <div
      class="login__menu-login-register"
    >
      <h5>
        Log out
      </h5>
    </div>
  </div>
</div>
`;

exports[`Login Menu should open the login form in the login menu 1`] = `
<div>
  <div
    class="LoginMenu__LoginMenuContainer-sc-1ec36x2-0 guOatf visible logged-out"
  >
    <div
      class="login__menu-login-register"
    >
      <h5
        data-testid="login-form-control"
      >
        Login
      </h5>
      <div
        class="LoginMenu__AuthFormContainer-sc-1ec36x2-1 bIHWLO visible"
      >
        <div
          class="Login__AuthFormContainer-sc-1ryrs0l-0 hBRKZY"
        >
          <form
            class="Login__Form-sc-1ryrs0l-1 NjHaG"
          >
            <div
              class="Login__FormInput-sc-1ryrs0l-2 iNdDlt"
            >
              <div
                class="Input__InputContainer-sc-1wyp7tv-0 eyXXuZ"
              >
                <input
                  class="invalid untouched "
                  data-testid="form-input"
                  id="userEmail"
                  placeholder="Email"
                  type="email"
                  value=""
                />
              </div>
            </div>
            <div
              class="Login__FormInput-sc-1ryrs0l-2 iNdDlt"
            >
              <div
                class="Input__InputContainer-sc-1wyp7tv-0 eyXXuZ"
              >
                <input
                  class="invalid untouched "
                  data-testid="form-input"
                  id="password"
                  placeholder="Password"
                  type="password"
                  value=""
                />
              </div>
            </div>
            <button
              class="Login__AuthButton-sc-1ryrs0l-3 emqpgl disabled"
              disabled=""
            >
              Login
            </button>
          </form>
        </div>
      </div>
      <a
        href="/signup"
      >
        <h5>
          Register
        </h5>
      </a>
    </div>
  </div>
</div>
`;
